[lldb/Commands] Alias `script` command to `scripting run`

medismailben opened the pull request:
This patch introduces a new top-level `scripting` command with an `run` sub-command, that basically replaces the `script` raw command.

To avoid breaking the `script` command usages, this patch also adds an `script` alias to the `scripting run` sub-command.

The reason behind this change is to have a top-level command that will cover scripting related subcommands.

jimingham commented on lldb/source/Commands/CommandObjectScripting.h:
```diff
@@ -9,14 +9,21 @@
+  ~CommandObjectMultiwordScripting() override;
+};
+
+class CommandObjectScriptingExecute : public CommandObjectRaw {
```
We're somewhat inconsistent about this, but we don't tend to put the subcommands of the multiword commands in the header file.  It's not necessary since they get added to the command in the .cpp file and aren't needed more globally.

jimingham commented on lldb/test/API/functionalities/abbreviation/TestAbbreviations.py:
```diff
@@ -80,7 +80,9 @@ def test_command_abbreviations_and_aliases(self):
         # Check a command that wants the raw input.
         command_interpreter.ResolveCommand(r"""sc print("\n\n\tHello!\n")""", result)
         self.assertTrue(result.Succeeded())
-        self.assertEqual(r"""script print("\n\n\tHello!\n")""", result.GetOutput())
```
Since you are explicitly making the default lldb treat `script` as an alias for script execute, it might be good to test that as well.  We're already testing `sc` just above.

jimingham reviewed (approved):
LGTM with a couple little nits.

kastiglione commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -518,6 +518,15 @@ void CommandInterpreter::Initialize() {
 
+  cmd_obj_sp = GetCommandSPExact("scripting execute");
+  if (cmd_obj_sp) {
+    AddAlias("sc", cmd_obj_sp);
+    AddAlias("scr", cmd_obj_sp);
+    AddAlias("scri", cmd_obj_sp);
+    AddAlias("scrip", cmd_obj_sp);
+    AddAlias("script", cmd_obj_sp);
```
this is unfortunate, do these pollute help output? if so, can we hide them?

is there no way to make a prefix match an alias, when it's shorter?

jimingham replied:
The problem is that if we have the command `scripting` and the alias `script` and we see `sc` we need to resolve that command ambiguity.  The only tool we have for ambiguous command resolution at present is "exact matches always win".

If we wanted to solve this w/o making all the exact matches we intend to win, we could add a ranking system to the commands and aliases, which bias ambiguous matches in favor of one or the other of the commands, or some similar system.  I worry that would get messy pretty quickly, but anyway, you'd need something like that.

medismailben replied:
I agree its not the most elegant solution but I'd prefer solving the partial matching issue in a follow-up.

JDevlieghere replied:
+1

kastiglione reviewed (approved):
In addition to the comment about unique prefixes, my only other comment is, have you considered `scripting run`?

jimingham commented:
> In addition to the comment about unique prefixes, my only other comment is, have you considered `scripting run`?

The problem with this command is that it either executes a script given to it on the command line, or it runs the embedded script interpreter...

medismailben committed:
[lldb/Commands] Alias `script` command to `scripting run`

This patch introduces a new top-level `scripting` command with an
`run` sub-command, that basically replaces the `script` raw command.

To avoid breaking the `script` command usages, this patch also adds an
`script` alias to the `scripting run` sub-command.

The reason behind this change is to have a top-level command that will
cover scripting related subcommands.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
> LLVM Buildbot has detected a new failure on builder `lldb-aarch64-windows` running on `linaro-armv8-windows-msvc-05` while building `lldb` at step 6 "test".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/141/builds/425
> 
> Here is the relevant piece of the build log for the reference:
> 
> ```
> Step 6 (test) failure: build (failure)
> ...
> 31.128 [1/2/288] Linking CXX executable tools\lldb\unittests\Thread\ThreadTests.exe
> 31.168 [1/1/289] Linking CXX executable tools\lldb\unittests\ValueObject\LLDBValueObjectTests.exe
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:57: note: using lit tools: C:\Users\tcwg\scoop\apps\git\2.39.0.windows.2\usr\bin
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:508: note: using clang: c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:508: note: using ld.lld: c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\ld.lld.exe
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:508: note: using lld-link: c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\lld-link.exe
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:508: note: using ld64.lld: c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\ld64.lld.exe
> llvm-lit.py: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\llvm\utils\lit\lit\llvm\config.py:508: note: using wasm-ld: c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\wasm-ld.exe
> 31.169 [0/1/289] Runnin-- Testing: 1983 tests, 2 workers --
> Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 
> FAIL: lldb-shell :: SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp (1616 of 1983)
> ******************** TEST 'lldb-shell :: SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp' FAILED ********************
> Exit Code: 1
> 
> Command Output (stdout):
> --
> # RUN: at line 19
> c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe --target=specify-a-target-or-use-a-_host-substitution -target x86_64-pc-linux -gdwarf-5 -gsplit-dwarf    -fdebug-types-section -gpubnames -c C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp -o C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.main.o
> # executed command: 'c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe' --target=specify-a-target-or-use-a-_host-substitution -target x86_64-pc-linux -gdwarf-5 -gsplit-dwarf -fdebug-types-section -gpubnames -c 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp' -o 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.main.o'
> # note: command had no output on stdout or stderr
> # RUN: at line 21
> c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe --target=specify-a-target-or-use-a-_host-substitution -target x86_64-pc-linux -gdwarf-5 -gsplit-dwarf -DVARIANT    -fdebug-types-section -gpubnames -c C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp -o C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.foo.o
> # executed command: 'c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe' --target=specify-a-target-or-use-a-_host-substitution -target x86_64-pc-linux -gdwarf-5 -gsplit-dwarf -DVARIANT -fdebug-types-section -gpubnames -c 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp' -o 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.foo.o'
> # note: command had no output on stdout or stderr
> # RUN: at line 23
> c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\ld.lld.exe C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.main.o C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.foo.o -o C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp
> # executed command: 'c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\ld.lld.exe' 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.main.o' 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.foo.o' -o 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp'
> # .---command stderr------------
> # | ld.lld: warning: cannot find entry symbol _start; not setting start address
> # `-----------------------------
> # RUN: at line 26
> rm -f C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.dwp
> # executed command: rm -f 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp.dwp'
> # note: command had no output on stdout or stderr
> # RUN: at line 27
> c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\lldb.exe --no-lldbinit -S C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/build/tools/lldb\test\Shell\lit-lldb-init-quiet    -o "type lookup IntegerType"    -o "type lookup FloatType"    -o "type lookup CustomType"    -b C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp | c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\filecheck.exe C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp --check-prefix=NODWP
> # executed command: 'c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\lldb.exe' --no-lldbinit -S 'C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/build/tools/lldb\test\Shell\lit-lldb-init-quiet' -o 'type lookup IntegerType' -o 'type lookup FloatType' -o 'type lookup CustomType' -b 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\tools\lldb\test\Shell\SymbolFile\DWARF\x86\Output\dwp-foreign-type-units.cpp.tmp'
> # note: command had no output on stdout or stderr
> # executed command: 'c:\users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\filecheck.exe' 'C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp' --check-prefix=NODWP
> # .---command stderr------------
> # | C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp:35:11: error: NODWP: expected string not found in input
> # | // NODWP: (lldb) type lookup FloatType
> # |           ^
> # | <stdin>:20:22: note: scanning from here
> # |  typedef unsigned int IntegerType;
> # |                      ^
> # | <stdin>:21:3: note: possible intended match here
> # |  typedef float FloatType;
> # |   ^
> ```

This is a flakey test that was already failing in previous builds:  https://lab.llvm.org/buildbot/#/builders/141/builds/423

kastiglione commented:
> The problem with this command is that it either executes a script given to it on the command line, or it runs the embedded script interpreter...

I don't see the problem. The original `script` does both of those things, and `run` works fits those two use case as well or better than `execute`, in my opinion.

medismailben commented:
> > The problem with this command is that it either executes a script given to it on the command line, or it runs the embedded script interpreter...
> 
> 
> 
> I don't see the problem. The original `script` does both of those things, and `run` fits those two use case as well or better than `execute`, in my opinion.

I don't know if you saw but I actually changed the command to be `scripting run` before landing this 😅

kastiglione commented:
@medismailben 👍 I saw, and also wanted to reply to Jim.

JDevlieghere commented:
> The reason behind this change is to have a top-level command that will cover scripting related subcommands.

Which other scripting subcommands do you envision, besides the [template](https://github.com/llvm/llvm-project/pull/97273) subcommand?

medismailben commented:
> The reason behind this change is to have a top-level command that will cover scripting related subcommands.

From @jimingham reply:

Which other scripting subcommands do you envision, besides the [template](https://github.com/llvm/llvm-project/pull/97273) subcommand? 

One that I want, powered by facilities we can add to the template, is that a class that implements a breakpoint callback should have a registry method (that's the part we can probably generate with the template instantiation) that adds callback class and help info to lldb.  Then you would do:

```
(lldb) scripting instances list
Breakpoint Callbacks:
  my_module.MyBreakpointCallback: 
      This does something really cool, if you pass the right value for the key "even_cooler", it does something even cooler.


```
And similar for all the other affordances.  We could do this by having `breakpoint callback implementation list` but I think having them listed here would make all the different affordances we have more discoverable.

BTW, I'm not proposing `instances` as the name, but I couldn't think of a better one off the top of my head.

Maybe `scripting callback list`?

It might also be handy to have:

(lldb) scripting modules list -l Python
MyFirstModule : /Users/Me/MyPythonModules/my_first_script.py

etc...

Which would list all the Python modules that we've loaded with "command script import" and where they came from.
